tests/gradients/test_spsa_gradient.py

```diff
@@ -0,0 +1,878 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests for the gradients.spsa_gradient module.
+"""
+import pytest
+
+from pennylane import numpy as np
+
+import pennylane as qml
+from pennylane.gradients import spsa_grad
+from pennylane.gradients.spsa_gradient import _rademacher_sampler
+from pennylane.devices import DefaultQubit
+from pennylane.operation import Observable, AnyWires
+
+
+def coordinate_sampler(indices, num_params, idx):
+    """Return a single canonical basis vector, corresponding
+    to the index ``indices[idx]``. This is a sequential coordinate sampler
+    that allows to exactly reproduce derivatives, instead of using SPSA in the
+    intended way."""
+    idx = idx % len(indices)
+    direction = np.zeros(num_params)
+    direction[indices[idx]] = 1.0
+    return direction
+
+
+class TestRademacherSampler:
+    @pytest.mark.parametrize(
+        "ids, num", [(list(range(5)), 5), ([0, 2, 4], 5), ([0], 1), ([2, 3], 5)]
+    )
+    def test_output_structure(self, ids, num):
+        ids_mask = np.zeros(num, dtype=bool)
+        ids_mask[ids] = True
+
+        for _ in range(5):
+            direction = _rademacher_sampler(ids, num)
+            assert direction.shape == (num,)
+            assert set(direction).issubset({0, -1, 1})
+            assert np.allclose(np.abs(direction)[ids_mask], 1)
+            assert np.allclose(direction[~ids_mask], 0)
+
+    @pytest.mark.parametrize("num_directions, tol", [(100, 0.1), (500, 5e-3)])
+    def test_convergence_single_par(self, num_directions, tol):

```

Comment:
Should it not be in the other class, I think we should keep it for Rademacher unit tests and not integration.

Reply:
Good point, thanks!

Reply:
I moved it and added a convergence test to the Rademacher test class.